### ISSUE TITLE
Function groups

### ISSUE THREAD

**Issue by ajcrites** [apex/apex#306]
Current directory structure for functions is `functions/{function,names}/src`. It would be useful to group functions together. My suggestion would be to have another optional directory that would specify the group such as:

    functions/{group1,group2}/{fna,fnb,fnc}/src

The use case for this is that it is simpler to do deployments for groups of functions that share parameters with each other but not with other functions in the same project.

    apex deploy group1 -e 'env_1=value'
    apex deploy group2 -e 'env_2=value'

This is just a suggestion since there are a number of ways to achieve this like having separate projects entirely or maintaining your own list of function groups, but I think this would be a nice, intuitive feature.

**Comment by tj** [comment]
HmmMmM. Personally I've been breaking groups into completely separate repos with their associated Terraform config etc but we can leave this open for feedback. For < v1.0 it would be unnecessary complications I think.

**Comment by cortesi** [comment]
Allowing function directories to nest would be really helpful to me. Directories could translate to lambda names like so:

    functions/a/b   -> project_a_b
    functions/a/b/c -> project_a_b_c

This would let one easily segregate prod/* and dev/* functions, duplicate groups with a cp -r, apply structure, etc. There are other ways to handle all of these with multiple projects, but this seems like a natural extension.

**Comment by tj** [comment]
I'd like to have envs as first-class concepts, but they'd be using the same functions (you could tweak behaviour with env vars etc). We should have that pretty soon since I'll need it right away too haha

**Comment by cortesi** [comment]
Look forward to seeing what you have in mind with envs. In my case at least, switching between entire accounts would sometimes be quite cumbersome and require a lot of duplicate infrastructure. For example, in one case I have a continuous stream of data I process with lambdas using S3 events. I deploy dev iterations in parallel to the production lambdas working on the same data stream, before promoting them to production. The same goes for API Gateway stages, where I typically have prod/dev running in parallel pointing at different sets of lambdas. It would be nice to have a consistent way of managing this.

**Comment by tj** [comment]
You can set up roles so there's a little switcher in the Console, it's not bad if you do that. I'm pretty opinionated when it comes to separate accounts for isolation, that part I may not budge from ;D. If it's worth having in production it's probably worth having a real staging env IMO. 

As far as canary deploys go that's something I do want: #7

**Comment by cortesi** [comment]
OK, I hope you don't mind if I keep pushing a bit here. Opinionated is good, but I'd like to convince you keep Apex open to work patterns like mine.

So, the way I work there are more than simply staging and production deploys. I have a lot of gray areas - for example, different versions of an API deployed simultaneously, running on the same infrastructure. Older API versions might be deprecated after a time, and development versions might be promoted to production after a time. Exactly the same goes for data ingest, where we have different versions of ingestion processes running alongside each other. This is not shoddy isolation: on the contrary, it's a well-managed process that lets us trial features and experiment carefully. Lambdas lend themselves particularly well to this approach, and Apex could help with just a wee bit more structure. 

The other aspect is that I don't use Terraform - I'm using CloudFormation. There are practical issues here. For instance, CloudFormation for reasons mysterious doesn't let you specify Role names in your templates. Instead, role names are auto-generated with a random component. This means that if one did create a duplicate CloudFormation infrastructure all the role ARNs would differ and one would have to cater for switching those in Apex when switching roles. At the moment I do this manually, and with different accounts maintaining this will quickly become a headache. CloudFormation can also take a bloody long time to set up resources - completing setup of one of our templates that involve CloudFront and Route53 can sometimes take an hour. Unfortunately, one does not simply spin up and tear down a substantial CloudFront infrastructure.

And +1 for canary deploys, which will simplify my test invocation hack substantially.

**Comment by tj** [comment]
Hmm I can see the API versioning use-case. As far as testing out functions which are promoted (canaries) I'd personally still use aliases (added a PR https://github.com/apex/apex/pull/312). I don't typically merge to master unless it's prod-ready, so I would probably keep the same structure but held in a different branch, deployed as `--alias canary` or similar. Any objections to that over copying?

Also valid :D haha. One of the big reasons I don't want to encourage multiple envs per account is just that it would complicate the Apex work flow I think, but if we can all agree on some tooling that doesn't do that I'm happy. Grouped functions is definitely abstract enough that I don't think it encourages that, however other things later like using API Gateway's stages do.

**Comment by cortesi** [comment]
Keeping API versions in branches will sometimes make sense, and sometimes not (I can go into why, if you're interested). That might be moot, though - if you agree that grouped functions (function hierarchies?) are light enough and have a range of use cases, we're golden. I'm happy to do a first draft of this and put a PR up for discussion, if you like (probably over this coming weekend).

Frankly, I'd like it if Apex remained agnostic about API Gateway, Terraform/Cloudfront,  isolation strategies and so forth. It's an excellent, lightweight system for managing lambdas, and I already have ways to handle these other pieces of my AWS infrastructure. I think the ability to specify Roles directly in the Apex configuration is perhaps as far as I want Apex to encroach (maybe your answer to that is "terraform", though).

**Comment by tj** [comment]
Hmmm I'm still a little on-edge about the feature, just because it adds another layer of docs and "UI" to specify groups etc, if we can do without that would be nice. What are the cases when you wouldn't have an old 1.x branch etc?

Yeah I've been using TF for roles and pretty much everything else, I think it's getting API Gateway support as well so we probably won't need to go there.

**Comment by ajcrites** [comment]
I was mostly interested in function groups because there is a lot of content that is shared between my function psuedo-groups that I have set up now, i.e. node modules, configuration files, and even the project.json. I could copy all of that between the two groups (in fact I do with a script) but it would be nice to have a lot of the crossover data and then deploy multiple functions at a time in groups. The reason for the separation would be different environment variables and different roles. I can cross these over, sure, but I'd rather not.

I just think the grouping is intuitive but if it's not or it's too much work ... or there is a better solution to groups of functions sharing environment variables within the same project ... maybe it's unnecessary.

**Comment by cortesi** [comment]
@tj This comes down to whether your Apex project can sensibly live in its own repo. When it can, I think API versioning with branches might make sense. In the project I'm trialling Apex with, this would be pain. The lambdas don't stand alone - they're part of a broader tool chain, static assets are copied into them on build, and they are tied closely to the rest of a pretty swift-moving project. We often make changes that affect affect the implementation but not the interface of all versions of the API, and we often (usually?) want to deploy everything at once.  I can imagine handling this with a separate repo and a set of scripts, but it would be needlessly baroque. API versioning is also only one use case here - there are other reasons to want to structure things a bit more. 

I think this could be a feature that doesn't add much complexity. For instance, I don't think introducing groups as a first class concept is necessary - instead, we can just naturally generate names for functions using their path. The main thing from the user's perspective would then be how functions are specified on the command-line. 

Say we had a directory structure like so:

```
functions
       +---> a
       |
       +---> x +-->  b
       |
       +---> y +-->  b
                |
                +-->  c
```

Now, ```apex deploy a``` would do exactly what it does today. ```apex deploy b``` would deploy project_x_b and project_y_b. To make specification more convenient, we can treat an argument containing path separators as a directory rather than a function name, in which case all contained functions are matched. So, ```apex deploy ./functions/y``` would deploy project_y_b and project_y_c, and ```apex deploy ./functions/y/b``` would do the obvious thing.

**Comment by tj** [comment]
@cortesi I like that idea actually, keeps it nice and simple conceptually.

@ajcrites I think for your case sub-projects almost make more sense, if it's getting into the world of having multiple levels of project.json-ish config, unless we took a similar approach to above, and just pick up project.json files wherever they happen to live in the tree.

As long as the rules are easy to grasp I'm totally down for stuff like this. I was toying with the idea of "decomposing" json into arbitrary files based on path https://github.com/tj/go-decompose, which is kind of similar to what @ajcrites is looking for.

I'll give this some more thought but I think I'm +1. It would be cool to treat function.json's as an extension of project.json in this "tree", maybe then it'll be more natural to place these files where-ever they are needed, cascading down.

**Comment by tj** [comment]
Though I guess really what utility does the tree structure provide? If we're only including the function itself in the zip, it doesn't really change anything.

**Comment by cortesi** [comment]
Cascading the settings would be a really nice bonus - being able to set the roles for entire groups of functions in one place and have that cascade up would be super handy. The main benefit for me is still organisational - with 40 lambdas spread over 3 API versions, I just need the structure.

**Comment by mthenw** [comment]
> apex deploy b would deploy project_x_b and project_y_b.

I think that this behaviour is unpredictable.

>  To make specification more convenient, we can treat an argument containing path separators as a directory rather than a function name

Makes more sense imho but I would treat ./functions as a root dir so:
- `apex deploy a` deploys `a`
- `apex deploy x/b` deploys `b` from `x` dir
- etc.

**Comment by cortesi** [comment]
> Makes more sense imho but I would treat ./functions as a root dir so:

The problem with that is that command-line path completion on the shell would no longer work - that's an important user convenience that we should retain.

I think avoiding the proxect_x_b/project_y_b behaviour would be OK - I can see how that might be confusing. Down the track, we could consider having a more powerful matching syntax based on Go's path globbing:

    apex deploy "functions/v1/*""
    apex deploy "**/a"

**Comment by tj** [comment]
groups would be really handy for viewing logs too actually :+1: now if I can only find a way to lower Lambda cost so I can actually afford to use it :D

**Comment by tj** [comment]
I was thinking we could just glob as an alternative. Not against grouping via dirs, but simplest for now would be to have `apex logs -f checks*` etc match `checks_{processor,queuer,reporter}` and so on

**Comment by tj** [comment]
Actually, globbing is potentially more powerful I think, you could do `apex logs -f *_reporter`, dirs wouldn't enable that as easily

**Comment by mthenw** [comment]
I was starting implementing this feature but I think we need to rethink how the functions are detected. Requiring `group.json` and `function.json` in nested functions provides lots of noise. Most of the time those file will be empty. I think that current inference feature guides to clean and minimalistic functions dir.

For now I'm removing from 0.9.0 milestone.

**Comment by tj** [comment]
wonder if we can figure out from stats how many people use function.json files, that might give us a good answer there actually. I have them in each dir personally, just because I like being explicit about memory/timeout, but our inference stuff is kinda cool
